redis_collections/dicts.py

```diff
@@ -122,70 +120,61 @@ def __getitem__(self, key):
         method with the key *key* as argument. The ``d[key]`` operation
         then returns or raises whatever is returned or raised by
         the ``__missing__(key)`` call if the key is not present.
-
-        .. note::
-            Due to implementation on Redis side, this method of retrieving
-            items is not very efficient. If possible, use :func:`get`.
         """
-        with self.redis.pipeline() as pipe:
-            pipe.hexists(self.key, key)
-            pipe.hget(self.key, key)
-            exists, value = pipe.execute()
+        key_hash, D = self._get_hash_dict(key, self.redis)
 
-        if not exists:
+        try:
+            value = D[key]
+        except KeyError:
             if hasattr(self, '__missing__'):
                 return self.__missing__(key)
-            raise KeyError(key)
-        return self._unpickle(value)
+            else:
+                raise
+
+        return value
 
     def __setitem__(self, key, value):
         """Set ``d[key]`` to *value*."""
-        value = self._pickle(value)
-        self.redis.hset(self.key, key, value)
+        key_hash, D = self._get_hash_dict(key, self.redis)
+        D[key] = value
 
-    def __delitem__(self, key):
-        """Remove ``d[key]`` from dictionary. Raises
-        a :func:`KeyError` if *key* is not in the map.
+        self.redis.hset(self.key, key_hash, self._pickle(D))
 
-        .. note::
-            Due to implementation on Redis side, this method of deleting
-            items is not very efficient. If possible, use :func:`discard`.
+    def __delitem__(self, key):
+        """Remove ``d[key]`` from dictionary.
+        Raises a :func:`KeyError` if *key* is not in the map.
         """
-        with self.redis.pipeline() as pipe:
-            pipe.hexists(self.key, key)
-            pipe.hdel(self.key, key)
-            exists, _ = pipe.execute()
+        key_hash, D = self._get_hash_dict(key, self.redis)
+        del D[key]
 
-        if not exists:
-            raise KeyError(key)
+        if D:
+            self.redis.hset(self.key, key_hash, self._pickle(D))
+        else:
+            self.redis.hdel(self.key, key_hash)
 
-    def discard(self, key):
```

Comment:
Zapped `discard` because it's just as (in)effecient as `__delitem__` now. I don't think the latter needed to be that inefficient anyway, since [HDEL](http://redis.io/commands/hdel) signals whether the field existed, but the point is moot now.